GIMP dialog windows misbehavior on multi-headed setup

This may be difficult to describe, but let's try. I work on a laptop with 2 external displays:

```text
[DP-1]
layout = 0,0
mode = 1920x1080@60.000000
scale = 1.000000
transform = normal

[eDP-1]
layout = 1920,0
mode = 1920x1080@120.000000
scale = 1.000000
transform = normal

[HDMI-A-1]
layout = 3840,0
mode = 1920x1080@60.000000
scale = 1.000000
transform = normal
```

If I open GIMP on `DP-1`, it works correctly on all workspaces. The problem appears on `eDP-1` and `HDMI-A-1`. I use the default 3x3 workspaces matrix. If I open GIMP on one of workspaces in the left column (1, 4 or 7), dialog popups are being moved to the left margin. On all other workspaces, they land somewhere off the screen. I use Waybar with the `wlr/taskbar` module. If I click on the button corresponding to the missing window, I'm being moved to the first left workspace in the same row. 

Sample steps to reproduce:

- launch GIMP on a display which coordinates are not 0,0 and workspace 2, 3, 5, 6, 8 or 9;
- open an image
- select Image / Canvas size
- change width or height, press Center

The dialog popup disappears from the screen.

I stole a spare monitor from another desk, but for some reason I'm unable to build:

```
$ ninja -C build
ninja: Entering directory `build'
[132/223] Compiling C++ object src/wayfire.p/view_xwayland.cpp.o
FAILED: src/wayfire.p/view_xwayland.cpp.o
c++ -Isrc/wayfire.p -Isrc -I../src -I. -I.. -I../src/api -Isubprojects/wf-config/include -I../subprojects/wf-config/include -Isubprojects/wf-utils -I../subprojects/wf-utils -Isubprojects/wf-touch -I../subprojects/wf-touch -I/usr/include/libdrm -I/usr/include/pixman-1 -I/usr/include/libpng16 -fdiagnostics-color=always -pipe -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -Wnon-virtual-dtor -Wextra -std=c++17 -g '-DWAYFIRE_VERSION="0.5.0-4d3d6c3 (" __DATE__ ", branch '"'"'HEAD'"'"')"' -DWLR_USE_UNSTABLE -Wno-unused-parameter -DHAS_ADDR2LINE=1 -MD -MQ src/wayfire.p/view_xwayland.cpp.o -MF src/wayfire.p/view_xwayland.cpp.o.d -o src/wayfire.p/view_xwayland.cpp.o -c ../src/view/xwayland.cpp
../src/view/xwayland.cpp: In lambda function:
../src/view/xwayland.cpp:364:24: error: ‘wlr_xwayland_minimize_event’ was not declared in this scope; did you mean ‘wlr_xwayland_resize_event’?
  364 |             auto ev = (wlr_xwayland_minimize_event*)data;
      |                        ^~~~~~~~~~~~~~~~~~~~~~~~~~~
      |                        wlr_xwayland_resize_event
../src/view/xwayland.cpp:364:52: error: expected primary-expression before ‘)’ token
  364 |             auto ev = (wlr_xwayland_minimize_event*)data;
      |                                                    ^
../src/view/xwayland.cpp: In member function ‘virtual void wayfire_xwayland_view::initialize()’:
../src/view/xwayland.cpp:406:49: error: ‘struct wlr_xwayland_surface::<unnamed>’ has no member named ‘request_minimize’; did you mean ‘request_maximize’?
  406 |         on_request_minimize.connect(&xw->events.request_minimize);
      |                                                 ^~~~~~~~~~~~~~~~
      |                                                 request_maximize
../src/view/xwayland.cpp: In member function ‘virtual void wayfire_xwayland_view::set_minimized(bool)’:
../src/view/xwayland.cpp:615:13: error: ‘wlr_xwayland_surface_set_minimized’ was not declared in this scope; did you mean ‘wlr_xwayland_surface_set_maximized’?
  615 |             wlr_xwayland_surface_set_minimized(xw, minimized);
      |             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |             wlr_xwayland_surface_set_maximized
[137/223] Compiling C++ object src/wayfire.p/output_plugin-loader.cpp.o
ninja: build stopped: subcommand failed.
```

Can't build it. Wrong `wlroots` version?

```
$ ninja -C build
ninja: Entering directory `build'
[30/62] Compiling C++ object src/wayfire.p/view_xwayland.cpp.o
FAILED: src/wayfire.p/view_xwayland.cpp.o
c++ -Isrc/wayfire.p -Isrc -I../src -I. -I.. -I../src/api -Isubprojects/wf-config/include -I../subprojects/wf-config/include -Isubprojects/wf-utils -I../subprojects/wf-utils -Isubprojects/wf-touch -I../subprojects/wf-touch -I/usr/include/libdrm -I/usr/include/pixman-1 -I/usr/include/libpng16 -fdiagnostics-color=always -pipe -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -Wnon-virtual-dtor -Wextra -std=c++17 -g '-DWAYFIRE_VERSION="0.5.0-f56a9b0 (" __DATE__ ", branch '"'"'HEAD'"'"')"' -DWLR_USE_UNSTABLE -Wno-unused-parameter -DHAS_ADDR2LINE=1 -MD -MQ src/wayfire.p/view_xwayland.cpp.o -MF src/wayfire.p/view_xwayland.cpp.o.d -o src/wayfire.p/view_xwayland.cpp.o -c ../src/view/xwayland.cpp
../src/view/xwayland.cpp: In lambda function:
../src/view/xwayland.cpp:358:24: error: ‘wlr_xwayland_minimize_event’ was not declared in this scope; did you mean ‘wlr_xwayland_resize_event’?
  358 |             auto ev = (wlr_xwayland_minimize_event*)data;
      |                        ^~~~~~~~~~~~~~~~~~~~~~~~~~~
      |                        wlr_xwayland_resize_event
../src/view/xwayland.cpp:358:52: error: expected primary-expression before ‘)’ token
  358 |             auto ev = (wlr_xwayland_minimize_event*)data;
      |                                                    ^
../src/view/xwayland.cpp: In member function ‘virtual void wayfire_xwayland_view::initialize()’:
../src/view/xwayland.cpp:400:49: error: ‘struct wlr_xwayland_surface::<unnamed>’ has no member named ‘request_minimize’; did you mean ‘request_maximize’?
  400 |         on_request_minimize.connect(&xw->events.request_minimize);
      |                                                 ^~~~~~~~~~~~~~~~
      |                                                 request_maximize
../src/view/xwayland.cpp: In member function ‘virtual void wayfire_xwayland_view::set_minimized(bool)’:
../src/view/xwayland.cpp:609:13: error: ‘wlr_xwayland_surface_set_minimized’ was not declared in this scope; did you mean ‘wlr_xwayland_surface_set_maximized’?
  609 |             wlr_xwayland_surface_set_minimized(xw, minimized);
      |             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |             wlr_xwayland_surface_set_maximized
[39/62] Compiling C++ object plugins/decor/libdecoration.so.p/deco-layout.cpp.o
ninja: build stopped: subcommand failed.
```

--- a/src/view/xwayland.cpp
+++ b/src/view/xwayland.cpp
@@ -193,28 +193,91 @@ class wayfire_xwayland_view_base : public wf::wlr_view_t
         wf::wlr_view_t::destroy();
     }
 
+    /* Translates geometry from X client configure requests to wayfire
+     * coordinate system. The X coordinate system treats all outputs
+     * as one big desktop, whereas wayfire treats the current workspace
+     * of an output as 0,0 and everything else relative to that. This
+     * means that we must take care when placing xwayland clients that
+     * request a configure after initial mapping, while not on the
+     * current workspace.
+     *
+     * @param output    The view's output
+     * @param ws_offset The view's workspace minus the current workspace
+     * @param geometry  The configure geometry as requested by the client
+     *
+     * @return Geometry with a position that is within the view's workarea.
+     * The workarea is the workspace where the view was initially mapped.
+     * Newly mapped views are placed on the current workspace.
+     */
+    wf::geometry_t translate_geometry_to_output(wf::output_t *output,
+        wf::point_t ws_offset,
+        wf::geometry_t g)
+    {
+        auto outputs = wf::get_core().output_layout->get_outputs();
+        auto og   = output->get_layout_geometry();
+        auto from = wf::get_core().output_layout->get_output_at(
+            g.x + g.width / 2 + og.x, g.y + g.height / 2 + og.y);
+        if (!from)
+        {
+            return g;
+        }
+
+        auto lg = from->get_layout_geometry();
+        g.x += (og.x - lg.x) + ws_offset.x * og.width;
+        g.y += (og.y - lg.y) + ws_offset.y * og.height;
+        if (!this->is_mapped())
+        {
+            g.x *= (float)og.width / lg.width;
+            g.y *= (float)og.height / lg.height;
+        }
+
+        return g;
+    }
+
     virtual void configure_request(wf::geometry_t configure_geometry)
     {
         /* Wayfire positions views relative to their output, but Xwayland
          * windows have a global positioning. So, we need to make sure that we
          * always transform between output-local coordinates and global
-         * coordinates */
-        if (get_output())
+         * coordinates. Additionally, when clients send a configure request
+         * after they have already been mapped, we keep the view on the
+         * workspace where its center point was from last configure, in
+         * case the current workspace is not where the view lives */
+        auto o = get_output();
+        if (o)
         {
-            auto current_workspace =
-                get_output()->workspace->get_current_workspace();
-            auto wsize = get_output()->workspace->get_workspace_grid_size();
-            auto og    = get_output()->get_layout_geometry();
+            auto wsg = o->workspace->get_workarea();
+            auto og  = o->get_layout_geometry();
+
             configure_geometry.x -= og.x;
             configure_geometry.y -= og.y;
 
-            /* Make sure views don't position themselves outside the desktop
-             * area, which is made up of all workspaces for an output */
-            wlr_box wsg{
-                -current_workspace.x * og.width,
-                -current_workspace.y * og.height,
-                wsize.width * og.width,
-                wsize.height * og.height};
+            auto view = this->self();
+            while (view->parent)
+            {
+                view = view->parent;
+            }
+
+            auto vg = view->get_wm_geometry();
+            auto current_workspace = o->workspace->get_current_workspace();
+            auto ws_offset = wf::point_t{0, 0};
+            if (vg.width && vg.height)
+            {
+                vg.x += current_workspace.x * og.width;
+                vg.y += current_workspace.y * og.height;
+                auto center =
+                    wf::point_t{vg.x + vg.width / 2, vg.y + vg.height / 2};
+                auto view_ws =
+                    wf::point_t{int(center.x / og.width), int(center.y / og.height)};
+                ws_offset.x = view_ws.x - current_workspace.x;
+                ws_offset.y = view_ws.y - current_workspace.y;
+                wsg.x += ws_offset.x * og.width;
+                wsg.y += ws_offset.y * og.height;
+            }
+
+            configure_geometry = translate_geometry_to_output(o, ws_offset,
+                configure_geometry);
+
             configure_geometry = wf::clamp(configure_geometry, wsg);
         }
 
@@ -464,12 +527,7 @@ class wayfire_xwayland_view : public wayfire_xwayland_view_base
 
         if (!this->tiled_edges && !xw->fullscreen)
         {
-            /* Make sure the view is visible on the current workspace
-             * on the current output. */
-            auto output_geometry     = get_output()->get_layout_geometry();
-            wlr_box current_geometry = {xw->x, xw->y, xw->width, xw->height};
-            current_geometry = wf::clamp(current_geometry, output_geometry);
-            configure_request(current_geometry);
+            configure_request({xw->x, xw->y, xw->width, xw->height});
         }
 
         wf::wlr_view_t::map(surface);

